gstreamer/src/query.rs

```diff
@@ -15,188 +15,188 @@ use std::ptr;
+            unsafe { from_glib_full(ffi::gst_query_new_formats()) }
+        )
     }
 
-    pub fn new_drain() -> Self {
+    pub fn new_buffering(fmt: ::Format) -> Buffering {
         assert_initialized_main_thread!();
-        unsafe { from_glib_full(ffi::gst_query_new_drain()) }
+        Buffering(
+            unsafe { from_glib_full(ffi::gst_query_new_buffering(fmt.to_glib())) }
+        )
     }
 
-    pub fn new_context(context_type: &str) -> Self {
+    pub fn new_custom(structure: ::Structure) -> Custom {
         assert_initialized_main_thread!();
-        unsafe { from_glib_full(ffi::gst_query_new_context(context_type.to_glib_none().0)) }
-    }
-}
-
-impl QueryRef {
-    pub fn get_structure(&self) -> Option<&StructureRef> {
-        unsafe {
-            let structure = ffi::gst_query_get_structure(self.as_mut_ptr());
-            if structure.is_null() {
-                None
-            } else {
-                Some(StructureRef::from_glib_borrow(structure))
+        Custom(
+            unsafe {
+                from_glib_full(ffi::gst_query_new_custom(
+                    ffi::GST_QUERY_CUSTOM,
+                    structure.into_ptr(),
+                ))
             }
-        }
+        )
     }
 
-    pub fn get_mut_structure(&mut self) -> &mut StructureRef {
-        unsafe {
-            let structure = ffi::gst_query_writable_structure(self.as_mut_ptr());
-            StructureRef::from_glib_borrow_mut(structure)
-        }
+    pub fn new_uri() -> Uri {
+        assert_initialized_main_thread!();
+        Uri(
+            unsafe { from_glib_full(ffi::gst_query_new_uri()) }
+        )
     }
 
-    pub fn is_downstream(&self) -> bool {
-        unsafe {
-            ((*self.as_ptr()).type_ as u32) & (ffi::GST_QUERY_TYPE_DOWNSTREAM.bits()) != 0
-        }
+    pub fn new_scheduling() -> Scheduling {
+        assert_initialized_main_thread!();
+        Scheduling(
+            unsafe { from_glib_full(ffi::gst_query_new_scheduling()) }
+        )
     }
 
-    pub fn is_upstream(&self) -> bool {
-        unsafe {
-            ((*self.as_ptr()).type_ as u32) & (ffi::GST_QUERY_TYPE_UPSTREAM.bits()) != 0
-        }
+    pub fn new_accept_caps(caps: &::Caps) -> AcceptCaps {
+        assert_initialized_main_thread!();
+        AcceptCaps(
+            unsafe { from_glib_full(ffi::gst_query_new_accept_caps(caps.as_mut_ptr())) }
+        )
     }
 
-    pub fn is_serialized(&self) -> bool {
-        unsafe {
-            ((*self.as_ptr()).type_ as u32) & (ffi::GST_QUERY_TYPE_SERIALIZED.bits()) != 0
-        }
+    pub fn new_caps<'a, P: Into<Option<&'a ::Caps>>>(filter: P) -> Caps {
+        assert_initialized_main_thread!();
+        let filter = filter.into();
+        Caps(
+            unsafe { from_glib_full(ffi::gst_query_new_caps(filter.to_glib_none().0)) }
+        )
     }
 
-    pub fn view(&self) -> QueryView<&Self> {
-        let type_ = unsafe { (*self.as_ptr()).type_ };
+    pub fn new_drain() -> Drain {
```

Comment:
That the constructors return the concrete types seems useful to me in any case. And what is needed for these is that they can be somehow converted to a `GstRc<QueryRef>`, and that you can borrow (possibly mutably, that's why the `QueryView` had a type parameter before) a `&QueryRef` from it to pass it around.

So `view` on `GstRc<QueryRef>` could return a immutable `QueryView` somehow that allows to get an immutable `Drain`, etc (so we need some more type parameters). And `view_mut` on `GstRc<>` would give you the mutable variants. And that type parameter could probably be `&mut QueryRef` or `&QueryRef` like before?

I think with that we would cover all use cases without changing everything and without going away from consistency with the other types.

Reply:
I can give this a try, but isn't this what the `QueryWrapper` does? :)

One thing I prefer in my proposal is that the composition chain is more linear:
`QueryWrapper` owns `QueryView` owns `ConcreteQuery` owns `QueryRef`.

If I'm not mistaken, in current architecture, we have:
`QueryRef` owns `QueryView` owns `ConcreteQuery` owns a `ref` or `ref mut` to the initial `QueryRef`.

I had hard times understanding why it was designed that way.

Reply:
It's not so much about "owning" but about having another data structure that borrows from the other and allows accessing it in specific ways.

Current architecture is that `GstRc<QueryRef>` owns the query. You can borrow a `QueryRef` from it, mutably or not depending on refcount. If you have a `&QueryRef`, then you view into it (possibly mutably) which gives you access to the concrete content of the query (and the `&QueryRef` again for being able to modify that.

I think what could be done is that the concrete query views are made generic, so that you can get them by looking at a `&QueryRef` (possibly mutably) and also have them independently and then *owning* a `GstRc<QueryRef>` to which then can be transformed, and allow borrowing the `&QueryRef` (and mut variant) from it.

I.e. we would have something like `ConcreteQueryView<GstRc<QueryRef>>>`, `ConcreteQueryView<&mut QueryRef>` and `ConcreteQueryView<&QueryRef>`. I believe the first is only ever needed when we directly created a query and know that it has refcount==1, so that does not need to distinguish mutability.

Reply:
In any case, you're on to something and we'll find a nice solution for it all :)